### PR TITLE
[WIP] Full Refactor

### PR DESCRIPTION
**Pull request by mt-jones**
Full refactor merging
PR https://github.com/rapidsai/cudf/pull/352
PR https://github.com/rapidsai/cudf/pull/318

and bug fixes.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -39,15 +39,15 @@ script:
   # check conda versions
   - conda list
   # build ligdf
```
Comment (kkraus14) [inline .travis.yml]:
```suggestion
  # build licudf
```

File: cpp/modules/tests/rmm/cudf_test_fixtures.h [hunk 2]
```diff
@@ -1,7 +1,8 @@
 #pragma once
 
 #include "gtest/gtest.h"
-#include "rmm.h"
+
```
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/modules/tests/rmm/cudf_test_fixtures.h]: Agree -- all tests use these fixtures.

File: cpp/modules/utilities/scan.cu [hunk 3]
```diff
@@ -1,8 +1,7 @@
-#include <gdf/gdf.h>
-#include <gdf/utils.h>
-#include <gdf/errorutils.h>
-
-#include "rmm.h"
+#include "cudf.h"
```
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/modules/utilities/scan.cu]: Or a parallel_algorithms module that also includes sorting? But then we are getting a deeper hierarchy...  Ideally we wouldn't need our own parallel algorithms and would just call Thrust.

**Comment by harrism** [comment]
BTW, I also agree with Jake's high-level comments (comm and modules naming).

File: cpp/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+option(HT_LEGACY_ALLOCATOR "Use the legacy allocator for hash tables" ON)
+if(HT_LEGACY_ALLOCATOR)
+    message(STATUS "Using legacy allocator for hash tables")
+    set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} --define-macro HT_LEGACY_ALLOCATOR")
+endif(HT_LEGACY_ALLOCATOR)
+
+if(CMAKE_BUILD_TYPE MATCHES Debug)
+    message(STATUS "Building with debugging flags")
+    set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -G")
```
[the remark this answers is not in the document]
Comment (mt-jones) [reply cpp/CMakeLists.txt]: CMake adds “-g” for you if you specify debug build.

**Comment by harrism** [comment]
Does it add it to CUDA flags?
________________________________

**Comment by aucahuasi** [comment]
Hi team,

1) Would it be possible to choose the compute capability for the build process of cudf/libgdf? I would use a cmake argument for this, in fact we already did this for our internal libgdf fork since we needed to point to specific gpu targets.
2) Could you please generate static lib versions for all the cudf/libgdf targets (e.g. libgdf, librmm)? Again we already did this for our fork but it would be a nice feature (our approach was to pass a cmake option ... something like BUILD_STATIC_LIBS)

I know you are making big changes to the cmake files, but if you are going to take a long time I can make a PR with the above changes ;)

cc @harrism @kkraus14

**Comment by aucahuasi** [comment]
Hi Mark, great! I will push a PR with these 2 enhancements.
thanks!

File: cpp/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+            thirdparty/cnmem/src/cnmem.cpp
+            )
+
+###################################################################################################
+# - build options ---------------------------------------------------------------------------------
+
+option(USE_NVTX "Build with NVTX support" ON)
+if(USE_NVTX)
+    message(STATUS "Using Nvidia Tools Extension")
+    find_library(NVTX_LIBRARY nvToolsExt PATH ENV LD_LIBRARY_PATH)
```
[the remark this answers is not in the document]
Comment (mt-jones) [reply cpp/CMakeLists.txt]: Removed the dependence on `LD_LIBRARY_PATH` explicitly with `CMAKE_CUDA_IMPLICIT_LINK_DIRECTORIES`

**Comment by mt-jones** [comment]
> Hi team,
> 
> 1. Would it be possible to choose the compute capability for the build process of cudf/libgdf? I would use a cmake argument for this, in fact we already did this for our internal libgdf fork since we needed to point to specific gpu targets.
> 2. Could you please generate static lib versions for all the cudf/libgdf targets (e.g. libgdf, librmm)? Again we already did this for our fork but it would be a nice feature (our approach was to pass a cmake option ... something like BUILD_STATIC_LIBS)
> 
> I know you are making big changes to the cmake files, but if you are going to take a long time I can make a PR with the above changes ;)
> 
> cc @harrism @kkraus14

Regarding (1), you can specify the device architectures via `-DCMAKE_CUDA_FLAGS="-gencode=arch=compute_61,code=sm_61"` which would _add_ the flags to `nvcc`. Not override them. A note on this: adding additional architectures would increase build times. Long term, it'd be nice to control the target architecture with an override like `CUDA_ARCH=Volta` or similar.

(2) should be simple enough, but I think we should prioritize getting this merged.

**Comment by harrism** [comment]
@mike-wendt The cpp/README.md is updated, however I've requested a couple of changes. But you can get started testing CI, I think.

File: README.md [hunk 6]
```diff
@@ -22,15 +22,19 @@ conda install -c numba -c conda-forge -c rapidsai -c defaults cudf=0.2.0
...
 Note: This conda installation only applies to Linux and Python versions 3.5/3.6.
 
-You can create and activate a development environment using the conda command:
+You can create and activate a development environment using the conda commands:
 
 ```bash
-conda env create --name cudf --file conda_environments/testing_py35.yml
-source activate cudf
+# create the conda environment (assuming in base `cudf` directory)
+$ conda env create --name cudf_dev --file conda_environments/dev_py35.yml
```
Comment (jrhemstad) [inline README.md]: This filepath to the yml is incorrect. I had to do:

```conda env create --name cudf_dev --file conda/environments/dev_py35.yml```

Comment (kkraus14) [reply README.md]: Resolved.

**Comment by kkraus14** [comment]
@harrism we'd want them in `python/cudf/tests` as that's standard dir structure and allows for testing without having to install the python library.

**Comment by harrism** [comment]
rerun tests

**Pull request merged**